zephyr/drivers/can/can_sccan.c

```diff
@@ -0,0 +1,1147 @@
+}
+
+static void sc_can_get_data_frame(const struct sc_can_cfg *config, struct can_frame *frame)
+{
+	uint32_t data0_reg;
+	uint32_t data1_reg;
+
+	/* RMR3/RMR4 must be read to clear the FIFO regardless of the dlc */
+
+	data0_reg = sys_read32(config->reg_addr + SCCAN_RMR3_OFFSET);
+	frame->data_32[0] = sys_cpu_to_be32(data0_reg);
+
+	data1_reg = sys_read32(config->reg_addr + SCCAN_RMR4_OFFSET);
+	frame->data_32[1] = sys_cpu_to_be32(data1_reg);
+
+	LOG_DBG("Receiving %d bytes. Id: 0x%x, ID type: %s %s", frame->dlc, frame->id,
+		(frame->flags & CAN_FRAME_IDE) != 0 ? "extended" : "standard",
+		(frame->flags & CAN_FRAME_RTR) != 0 ? ", RTR frame" : "");
+}
+
+static void sc_can_tx_done(const struct device *dev, int status)
+{
+	const struct sc_can_cfg *config = dev->config;
+	struct sc_can_data *data = dev->data;
+	can_tx_callback_t callback;
+
+	callback = data->tx_cb_data_list[data->tx_tail].tx_cb;
+	if (callback == NULL) {
+		LOG_ERR("TX callback is not registerd.");
+		return;
+	}
+
+	callback(dev, status, data->tx_cb_data_list[data->tx_tail].tx_cb_arg);
+	data->tx_cb_data_list[data->tx_tail].tx_cb = NULL;
+
+	data->tx_tail++;
+	if (data->tx_tail >= config->tx_fifo_depth) {
+		data->tx_tail = 0;
+	}
+
+	return;
+}
+
+static inline void sc_can_enable_acceptance_filter(const struct sc_can_cfg *config, int filter_id)
+{
+	sys_set_bits(config->reg_addr + SCCAN_AFER_OFFSET, BIT(filter_id));
+}
+
+static void sc_can_set_acceptance_filter(const struct sc_can_cfg *config, int filter_id,
+					 const struct can_filter *filter)
+{
+	uint32_t mask_reg = 0;
+	uint32_t value_reg = 0;
+	bool extended = filter->flags & CAN_FILTER_IDE;
+	bool rtr = filter->flags & CAN_FILTER_RTR;
+
+	if (filter->flags & CAN_FILTER_IDE) {
+		mask_reg = (SCCAN_AFID_EX1(filter->mask) | SCCAN_AFSRTR(1) | SCCAN_AFIDE(extended) |
+			    SCCAN_AFID_EX2(filter->mask) | SCCAN_AFERTR(rtr));
+		value_reg = (SCCAN_AFID_EX1(filter->id) | SCCAN_AFSRTR(1) | SCCAN_AFIDE(extended) |
+			     SCCAN_AFID_EX2(filter->id) | SCCAN_AFERTR(rtr));
+	} else {
+		mask_reg = (SCCAN_TXID1(filter->mask) | SCCAN_AFSRTR(rtr) | SCCAN_AFIDE(extended));
+		value_reg = (SCCAN_TXID1(filter->id) | SCCAN_AFSRTR(rtr) | SCCAN_AFIDE(extended));
+	}
+
+	sys_write32(mask_reg, config->reg_addr + SCCAN_AFIMR1_OFFSET + (filter_id * 0x20));
+	sys_write32(value_reg, config->reg_addr + SCCAN_AFIVR1_OFFSET + (filter_id * 0x20));
+
+	return;
+}
+
+static void sc_can_add_acceptance_filter(const struct sc_can_cfg *config, int filter_id,
+					 const struct can_filter *filter)
+{
+	sc_can_enable_acceptance_filter(config, filter_id);
+
+	sc_can_set_acceptance_filter(config, filter_id, filter);
+}
+
+static void sc_can_rx_cb(const struct device *dev, struct can_frame *frame)
+{
+	const struct sc_can_cfg *config = dev->config;
+	struct sc_can_data *data = dev->data;
+	uint8_t filter_id;
+
+	/* In this driver, called the all RX callback that matched registered filter */
```

Comment:
I don't really get what you're trying to say. Are you trying to say "Call all RX callback functions that match the registered filter." ?